Updating behavior to not continue after an authenticator has set the response

weaverryan opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | https://github.com/symfony/symfony/pull/14673/files#r40492765 |
| License | MIT |
| Doc PR | n/a |

This mirrors the behavior in core: _if_ a listener sets a response (on success or failure),
then the other listeners are not called. But if a response is _not_ set
(which is sometimes the case for success, like in BasicAuthenticationListener),
then the other listeners are called, and can even fail.

It's all a bit of an edge-case, as only one authenticator (like authentication listener) would normally be doing any work on a request, but I think matching the other listeners (since I'm not aware of anyone having issues with its behavior) is best.

weaverryan committed:
Updating behavior to not continue after an authenticator has set the response

This mirrors the behavior in core: *if* a listener sets a response (on success or failure),
then the other listeners are not called. But if a response is *not* set
(which is sometimes the case for success, like in BasicAuthenticationListener),
then the other listeners are called, and can even fail.

stof commented on src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php:
```diff
@@ -75,6 +75,12 @@ public function handle(GetResponseEvent $event)
             $this->executeGuardAuthenticator($uniqueGuardKey, $guardAuthenticator, $event);
+
+            if ($event->hasResponse()) {
+                $this->logger->info(sprintf('The "%s" authenticator set the response. Any later authenticator will not be called', get_class($guardAuthenticator)));
```
this looks like a debug message, not an info one IMO. You only care about it if you need to debug something going wrong.

weaverryan replied:
That's a good point - we use `info()` everywhere in the other security listeners already though.

stof replied:
and part of them don't make sense at the info level IMO.

weaverryan committed:
Making all "debug" messages use the debug router

Only the "auth success" and "auth failed" messages remain at info. That's
consistent with AbstractAuthenticationListener

weaverryan commented on src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php:
```diff
@@ -66,7 +66,7 @@ public function __construct(GuardAuthenticatorHandler $guardHandler, Authenticat
     {
         if (null !== $this->logger) {
-            $this->logger->info('Checking for guard authentication credentials.', array('firewall_key' => $this->providerKey, 'authenticators' => count($this->guardAuthenticators)));
+            $this->logger->debug('Checking for guard authentication credentials.', array('firewall_key' => $this->providerKey, 'authenticators' => count($this->guardAuthenticators)));
```
This is unrelated to this PR, but it's so minor I snuck it in here. As Stof mentioned, most of these messages are really debug messages. I've only left the "auth success" and "auth failed" messages as `info()`, which is consistent with AbstractAuthenticationListener.

stof commented:
:+1: 

Status: reviewed

xabbuh commented:
:+1:

fabpot commented:
Thank you @weaverryan.
